### PR TITLE
feat:[SSCA-939]: Added logical operators in component version filter

### PR DESCRIPTION
**Pull request by arpit-harness**
## Describe your changes
Adding logical operator [ =, !=, >, >=, <, <= ]. These logical operator work on normalised component version fields [MajorVersion, MinorVersion, PatchVersion]. User can input a version string in formats [a.b.c, a.b, a] with operator selected in the operator field.

Screenshot 2023-11-24 at 10 55 20 AM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)

### CODE REVIEW

**Review by mankrit-singh — DISMISSED** [review]

**Comment by arpit-harness** [comment]
trigger uts

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java [hunk 1]
```diff
@@ -114,6 +113,105 @@ public Criteria getComponentCriteria(List<ComponentFilter> componentFilter) {
...
+    String fieldName = componentFilterToFieldNameMap.get(filter.getFieldName());
+    switch (filter.getOperator()) {
+      case EQUALS:
+        return Criteria.where(fieldName).is(filter.getValue());
+      case CONTAINS:
+        return Criteria.where(fieldName).regex(filter.getValue());
+      case STARTSWITH:
+        return Criteria.where(fieldName).regex(Pattern.compile("^".concat(filter.getValue())));
+      default:
+        throw new InvalidRequestException("Invalid component filter operator");
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:130]: nit: have better error message saying component name filter does not support _ operator.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java [hunk 2]
```diff
@@ -114,6 +113,105 @@ public Criteria getComponentCriteria(List<ComponentFilter> componentFilter) {
...
+    }
+  }
+
+  private Criteria getComponentVersionFilterCriteria(ComponentFilter filter) {
+    if (filter.getFieldName() != ComponentFilter.FieldNameEnum.COMPONENTVERSION) {
+      return new Criteria();
+    }
+    List<Integer> versions = new VersionField().getVersion(filter.getValue());
+    if (versions.size() != 3 || versions.get(0) == -1) {
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:139]: versions.size() != 3
Ask product if customers can give just version > 1
Also, ask product about what versions format will we support?
And if version is not in proper format, should we throw error? 
Get the correct error messages as well

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java [hunk 3]
```diff
@@ -114,6 +113,105 @@ public Criteria getComponentCriteria(List<ComponentFilter> componentFilter) {
     }
     return componentCriteria;
   }
+
+  private Criteria getComponentNameFilterCriteria(ComponentFilter filter) {
+    if (filter.getFieldName() != ComponentFilter.FieldNameEnum.COMPONENTNAME) {
```
Comment (dhruv-harness) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:118]: is this required, we are already checking this condition before calling this function

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java [hunk 4]
```diff
@@ -114,6 +113,105 @@ public Criteria getComponentCriteria(List<ComponentFilter> componentFilter) {
...
+        return Criteria.where(fieldName).regex(filter.getValue());
+      case STARTSWITH:
+        return Criteria.where(fieldName).regex(Pattern.compile("^".concat(filter.getValue())));
+      default:
+        throw new InvalidRequestException("Invalid component filter operator");
+    }
+  }
+
+  private Criteria getComponentVersionFilterCriteria(ComponentFilter filter) {
+    if (filter.getFieldName() != ComponentFilter.FieldNameEnum.COMPONENTVERSION) {
```
Comment (dhruv-harness) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:135]: same as above, is this check required again

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java [hunk 5]
```diff
@@ -114,6 +113,105 @@ public Criteria getComponentCriteria(List<ComponentFilter> componentFilter) {
...
+      return new Criteria();
+    }
+    List<Integer> versions = new VersionField().getVersion(filter.getValue());
```
Comment (dhruv-harness) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:138]: we can break down the impl for each operator in a separate function, this function has become too long

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)* status: 3 file(s) to review. 1 change requests. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java` line 174 at r1](https://reviewable.io/reviews/harness/harness-core/55917#-Njzqki53kof2wtaADNM:-Njzqki53kof2wtaADNN:by8syis) ([raw file](https://github.com/harness/harness-core/blob/a559995a974680eaa6c9120bb02ab72024994dc4/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java#L174)):*
> ```Java
>       majorCriteria = getNonPatchVersionCriteria(versions, filter.getOperator(), 1);
>     }
>     return new Criteria().orOperator(patchCriteria, minorCriteria, majorCriteria);
> ```

Why its an or operation ? why need 3 Criteria ? we can do it with one. 
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java` line 177 at r1](https://reviewable.io/reviews/harness/harness-core/55917#-NjzpFaK5aZ-_aQn39rv:-NjzpFaK5aZ-_aQn39rw:b-gdjcth) ([raw file](https://github.com/harness/harness-core/blob/a559995a974680eaa6c9120bb02ab72024994dc4/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java#L177)):*
> ```Java
>   }
> 
>   private Criteria getPatchVersionCriteria(List<Integer> versions, Operator operator) {
> ```

Why can't we handle it in generic way ? we need to map operator with field & condition.

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)* status: :shipit: complete! ✓

**Comment by arpit-harness** [comment]
trigger uts

**Comment by arpit-harness** [comment]
trigger uts

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java (hunk 5 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:134]: Seperated Equals and NotEquals cases from the rest.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java (hunk 4 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java]: Done

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java (hunk 3 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java]: Done.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java (hunk 2 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:135]: Added details on #eng-ssca channal.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java (hunk 1 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java:128]: Updated message.

**Review by arpit-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)* status: 2 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java` line 174 at r1](https://reviewable.io/reviews/harness/harness-core/55917#-Njzqki53kof2wtaADNM:-Nk-SXsA6FRpZ8rngVe3:b-efgaij) ([raw file](https://github.com/harness/harness-core/blob/1924e4977331be294ee4d74591a80019ed6f667d/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java#L174)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

Why its an or operation ? why need 3 Criteria ? we can do it with one.
</blockquote></details>

Following same as what we do in enforcement, for example if you say all versions greater than 1.2.3 this will result in three conditions.
1. major = 1, minor =2, patch >3
2. major =1, minor >2
3. major >1
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java` line 177 at r1](https://reviewable.io/reviews/harness/harness-core/55917#-NjzpFaK5aZ-_aQn39rv:-Nk-SxM-CMbu_U8dMwEr:b-v8wa0v) ([raw file](https://github.com/harness/harness-core/blob/1924e4977331be294ee4d74591a80019ed6f667d/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java#L177)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

Why can't we handle it in generic way ? we need to map operator with field & condition.
</blockquote></details>

We would need to map with functions and that too of two different types. This feels like a better approach in this scenario.

**Review by dhruv-harness — APPROVED** [review]

**Comment by arpit-harness** [comment]
trigger uts

**Review by inderschera — DISMISSED** [review]

**Review by dhruv-harness — DISMISSED** [review]

**Review by mankrit-singh — APPROVED** [review]

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — APPROVED** [review]
> Reviewed 1 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)* status: 6. unresolved discussions. 

___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java` line 177 at r1](https://reviewable.io/reviews/harness/harness-core/55917#-NjzpFaK5aZ-_aQn39rv:-Nk-nUFfC2DUzMBQrS6b:b6qh7jd) ([raw file](https://github.com/harness/harness-core/blob/1924e4977331be294ee4d74591a80019ed6f667d/ssca-manager/service/src/main/java/io/harness/ssca/services/NormalisedSbomComponentServiceImpl.java#L177)):*
<details><summary><i>Previously, arpit-harness wrote…</i></summary><blockquote>

We would need to map with functions and that too of two different types. This feels like a better approach in this scenario.
</blockquote></details>

Lets discuss this

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55917)* status: 6. unresolved discussions.

**Review by inderschera — APPROVED** [review]

**Review by mankrit-singh — APPROVED** [review]

**Pull request merged**